llvm/lib/Target/AMDGPU/AMDGPU.td

```diff
@@ -227,6 +227,22 @@ def FeatureLdsBranchVmemWARHazard : SubtargetFeature<"lds-branch-vmem-war-hazard
   "Switching between LDS and VMEM-tex not waiting VM_VSRC=0"
 >;
 
+class FeatureMaxHardClauseLength<int size> : SubtargetFeature<
+  "max-hard-clause-length-"#size,
+  "MaxHardClauseLength",
+  !cast<string>(size),
+  "Maximum number of instructions in an explicit S_CLAUSE is "#size
+>;
+
+def FeatureMaxHardClauseLength32 : FeatureMaxHardClauseLength<32>;
+def FeatureMaxHardClauseLength63 : FeatureMaxHardClauseLength<63>;
+def FeatureNoHardClauses : SubtargetFeature<
```

Comment:
What's `FeatureNoHardClauses` for? It seems unused by the rest of the patch.

Reply:
OK, if you want to keep it can you add a test that compiling for GFX10+ with an explicit "no-hard-clauses" attribute has the desired effect?

Reply:
Yeah, good point, will do